README
minismokebox
============

POE-based framework for automated CPAN Testing.

Supports multiple backends, including:

	- CPAN::YACSmoke
	- CPANPLUS::YACSmoke
	- CPAN::Reporter

Installation:
=============

The minimum version of perl supported is 5.6. This does not preclude CPAN testing of 
perls less than this, as the testing framework and the testing targets are two distinct
environments.

The included test-suite requires the following modules:

	Test::More 0.47
	File::Spec 0
	Test::POE::Server::TCP 0.12

The following modules are required:

	Config::Tiny => 2.12
	File::Spec => 0
	Pod::Usage => 1.00
	POE => 1.0001
	POE::Component::SmokeBox => 0.06
	POE::Component::SmokeBox::Recent => 1.00;
	POE::Component::SmokeBox::Dists => 0.08;

The recommended method of installation as always is to use either CPAN or CPANPLUS 
so that the dependencies are automagically dealt with for you and you have the option
of sending test reports during installation >:)

Status:
=======

Currently a testing release. Feel free to have a look, try it out, etc.

Any bug reports, feature requests, send an email to bug-minismokebox@rt.cpan.org


Changes
==============================
0.58 2013-09-29 18:40:04 +0100
==============================

  commit 66cacc633d6f7aff842c857e63f74ce633f10f41
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Sun Sep 29 18:36:49 2013 +0100
  
    Add support for osvers in the plugins 

==============================
0.56 2013-04-13 16:47:12 +0100
==============================

  commit 04c30ac32c66c8876b7c6362d4d6e055d23ca8a4
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Sat Apr 13 16:44:05 2013 +0100
  
    Require unless and use it. Add minismokebox.pm and move docs to there 

==============================
0.54 2011-06-02 23:49:06 +0100
==============================

  commit 2efa69f765fd2ab366e3c2c41620db04017d2f92
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Thu Jun 2 23:04:07 2011 +0100
  
    Fix a blunder with the previous release. Needed to affect which loop
    was loaded in the main module. 

==============================
0.52 2011-06-02 20:01:26 +0100
==============================

  commit 3810b1a7f17527f79f86621fb95c32ebff259f2b
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Thu Jun 2 20:01:26 2011 +0100
  
    Use POE::Loop::Event on MacOSX (darwin) because of problems with
    POE::XS::Loop::Poll

    Reported by Nigel Horne [rt.cpan.org #68608] 

==============================
0.50 2011-06-01 18:18:25 +0100
==============================

  commit 4002a44e81688985b4816b385e0866c613ef1e4a
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Wed Jun 1 18:18:25 2011 +0100
  
    Fix the autogenerated Changes file. No other changes (haha). 

==============================
0.48 2011-05-27 11:06:27 +0100
==============================

  commit 073582310cf72e870f58e34e0c9af9c39fec47c4
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Fri May 27 11:06:27 2011 +0100
  
    Bump version and add automated Changes generation

  commit 532283179bc014f12b2158bb09a8d30b311101ca
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Fri May 27 10:10:27 2011 +0100
  
    Add new --perlenv switch to pass PERL5LIB through to smoker process
    environment

  commit e11d5be66bbb7de21ddc144755c10d7e0d82f64a
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Fri May 27 10:01:35 2011 +0100
  
    Make the environment test actually test something

  commit c8c7a7a62998e8fbbd6cb0d155850f7dfab9a781
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Thu May 26 21:39:53 2011 +0100
  
    Make a better choice of POE event loops dependent on OS 

==============================
0.46 2011-05-25 22:08:49 +0100
==============================

  commit 541e523adb12c8fb6474bc8fab834e926fa309f5
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Wed May 25 22:08:49 2011 +0100
  
    Bump version and update Changes

  commit 49164f175930d9aa084dd5da4d9f4bc346ca067b
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Wed May 25 21:43:32 2011 +0100
  
    Remove trailing whitespace

  commit 022b10c28f9863f7c88f2d6337742f41aa2ee679
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Wed May 25 21:39:31 2011 +0100
  
    Added configuration options to pass custom env vars to the Backend
    process

  commit 68441b8574ce806ca62991efe36352a26aa41773
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Fri May 20 19:40:42 2011 +0100
  
    Account for potential blank lines or no data in the timestamp file 

==============================
0.44 2011-03-08 11:34:43 +0000
==============================

  commit 971cc602bb1bcf2bfe81e78130545b6d60ec8d5f
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Tue Mar 8 11:34:43 2011 +0000
  
    Update Changes

  commit 5d24d75f27cc81bef8b494dd8ae6a107afa4e17b
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Tue Mar 8 11:33:44 2011 +0000
  
    Update required version of POE::Filter::HTTP::Parser 

==============================
0.42 2011-02-10 21:53:27 +0000
==============================

  commit 3cff339bf5906fca466b6df1a0070c16e37a101e
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Thu Feb 10 21:53:27 2011 +0000
  
    Bump version to 0.42 and update Changes file

  commit 286f6618974b520b1da7368ecb08c0c4b749a27d
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Thu Feb 10 21:49:36 2011 +0000
  
    Make noepoch option actually do what it is supposed to do and work
    from the configuration file 

==============================
0.40 2010-12-09 20:01:31 +0000
==============================

  commit ab4096bf64a5e85be3218cb1af96a48f5a16582b
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Thu Dec 9 20:01:31 2010 +0000
  
    Finally get rid of the USE_SIGCHLD redefined warnings, add 'random'
    option to smoke random dists from CPAN 

==============================
0.38 2010-11-26 12:49:10 +0000
==============================

  commit 42e132b164bbcf85b9eb0bab5c21db2f789d0666
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Fri Nov 26 12:49:10 2010 +0000
  
    Bump version, update Changes

  commit 0ab5c078549081f8185cc58c37be64ecbf224485
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Fri Nov 26 12:48:07 2010 +0000
  
    Add noepoch option to explicitly disable the most recent uploads
    smoking based on when the smoker was last started. 

==============================
0.36 2010-11-16 22:14:36 +0000
==============================

  commit d8e01a54e21dbb13613a9dc6f243df4eb21f41c2
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Tue Nov 16 22:14:36 2010 +0000
  
    Remove debugging code.

  commit e114f0432d5bde39c52745609cb50b3142ddeb8a
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date:   Tue Nov 16 22:11:23 2010 +0000
  
    Release updates


lib/App/SmokeBox/Mini.pm
package App::SmokeBox::Mini;

use strict;
use warnings;
use Pod::Usage;
use Config::Tiny;
use File::Spec;
use Cwd;
use Getopt::Long;
use Time::Duration qw(duration_exact);
use POE;
use POE::Component::SmokeBox;
use POE::Component::SmokeBox::Smoker;
use POE::Component::SmokeBox::Job;
use POE::Component::SmokeBox::Dists;
use POE::Component::SmokeBox::Recent;

use vars qw($VERSION);

use constant CPANURL => 'ftp://ftp.funet.fi/pub/CPAN/';

$VERSION = '0.14';

$ENV{PERL5_MINISMOKEBOX} = $VERSION;

sub _smokebox_dir {
  return $ENV{PERL5_SMOKEBOX_DIR} 
     if  exists $ENV{PERL5_SMOKEBOX_DIR} 
     && defined $ENV{PERL5_SMOKEBOX_DIR};

  my @os_home_envs = qw( APPDATA HOME USERPROFILE WINDIR SYS$LOGIN );

  for my $env ( @os_home_envs ) {
      next unless exists $ENV{ $env };
      next unless defined $ENV{ $env } && length $ENV{ $env };
      return $ENV{ $env } if -d $ENV{ $env };
  }

  return cwd();
}

sub _read_config {
  my $smokebox_dir = File::Spec->catdir( _smokebox_dir(), '.smokebox' );
  return unless -d $smokebox_dir;
  my $conf_file = File::Spec->catfile( $smokebox_dir, 'minismokebox' );
  return unless -e $conf_file;
  my $Config = Config::Tiny->read( $conf_file );
  if ( defined $Config->{_} ) {
	return map { $_, $Config->{_}->{$_} } grep { exists $Config->{_}->{$_} }
		qw(debug perl indices recent backend url);
  }
  return;
}

sub _get_jobs_from_file {
  my $jobs = shift || return;
  unless ( open JOBS, "< $jobs" ) {
     warn "Could not open '$jobs' '$!'\n";
     return;
  }
  my @jobs;
  while (<JOBS>) {
    chomp;
    push @jobs, $_;
  }
  close JOBS;
  return @jobs;
}

sub _display_version {
  print "minismokebox version ", $VERSION, 
    ", powered by POE::Component::SmokeBox ", POE::Component::SmokeBox->VERSION, "\n\n";
  print <<EOF;
Copyright (C) 2009 Chris 'BinGOs' Williams
This module may be used, modified, and distributed under the same terms as Perl itself. 
Please see the license that came with your Perl distribution for details.
EOF
  exit;
}

sub run {
  my $package = shift;
  my %config = _read_config();
  my $version;
  GetOptions(
    "h|help"      => sub { pod2usage(1); },
    "v|version"   => sub { $version = 1 },
    "d|debug"     => \$config{debug},
    "p|perl=s" 	  => \$config{perl},
    "i|indices"   => \$config{indices},
    "r|recent"    => \$config{recent},
    "j|jobs=s"    => \$config{jobs},
    "b|backend=s" => \$config{backend},
    "a|author=s"  => \$config{author},
    "D|package=s" => \$config{package},
    "P|phalanx"   => \$config{phalanx},
    "u|url=s"	  => \$config{url},
    "R|reverse"   => \$config{reverse},
  ) or pod2usage(2);

  _display_version() if $version;

  $config{perl} = $^X unless $config{perl} and -e $config{perl};
  $ENV{PERL5_SMOKEBOX_DEBUG} = 1 if $config{debug};
  $ENV{AUTOMATED_TESTING} = 1;   # We need this because some backends do not set it.
  $ENV{PERL_MM_USE_DEFAULT} = 1; # And this.
  $ENV{PERL_EXTUTILS_AUTOINSTALL} = '--defaultdeps'; # Got this from CPAN::Reporter::Smoker. Cheers, xdg!

  if ( $config{jobs} and -e $config{jobs} ) {
     my @jobs = _get_jobs_from_file( $config{jobs} );
     $config{jobs} = \@jobs if scalar @jobs;
  }

  print "Running minismokebox with options:\n";
  printf("%-20s %s\n", $_, $config{$_}) 
	for grep { defined $config{$_} } qw(debug indices perl jobs backend author package phalanx reverse url);

  my $self = bless \%config, $package;

  $self->{sbox} = POE::Component::SmokeBox->spawn( 
	smokers => [
	   POE::Component::SmokeBox::Smoker->new(
		perl => $self->{perl},
	   ),
	],
  );

  $self->{session_id} = POE::Session->create(
	object_states => [
	   $self => { recent => '_submission', dists => '_submission', },
	   $self => [qw(_start _stop _check _indices _smoke _search)],
	],
	heap => $self,
  )->ID();

  $poe_kernel->run();
  return 1;
}

sub _start {
  my ($kernel,$self) = @_[KERNEL,OBJECT];
  $self->{session_id} = $_[SESSION]->ID();
  # Run a check to make sure the backend exists in the designated perl
  $kernel->post( $self->{sbox}->session_id(), 'submit', event => '_check', job => 
     POE::Component::SmokeBox::Job->new(
	( $self->{backend} ? ( type => $self->{backend} ) : () ),
	command => 'check',
     ),
  );
  $self->{stats} = {
	started => time(),
	totaljobs => 0,
	avg_run => 0,
	min_run => 0,
	max_run => 0,
	_sum => 0,
	idle => 0,
	excess => 0,
  };
  return;
}

sub _stop {
  my ($kernel,$self) = @_[KERNEL,OBJECT];
  $kernel->call( $self->{sbox}->session_id(), 'shutdown' );
  my $finish = time();
  my $cumulative = duration_exact( $finish - $self->{stats}->{started} );
  my @stats = map { $self->{stats}->{$_} } qw(totaljobs idle excess avg_run min_run max_run);
  $stats[$_] = duration_exact( $stats[$_] ) for 3 .. 5;
  print "minismokebox started at: \t", scalar localtime($self->{stats}->{started}), "\n";
  print "minismokebox finished at: \t", scalar localtime($finish), "\n";
  print "minismokebox ran for: \t", $cumulative, "\n";
  print "minismokebox tot jobs:\t", $stats[0], "\n";
  print "minismokebox idle kills:\t", $stats[1], "\n" if $stats[1];
  print "minismokebox excess kills:\t", $stats[2], "\n" if $stats[2];
  print "minismokebox avg run: \t", $stats[3], "\n";
  print "minismokebox min run: \t", $stats[4], "\n";
  print "minismokebox max run: \t", $stats[5], "\n";
  return;
}

sub _check {
  my ($kernel,$self,$data) = @_[KERNEL,OBJECT,ARG0];
  my ($result) = $data->{result}->results;
  unless ( $result->{status} == 0 ) {
     my $backend = $self->{backend} || 'CPANPLUS::YACSmoke';
     warn "The specified perl '$self->{perl}' does not have backend '$backend' installed, aborting\n";
     return;
  }
  if ( $self->{indices} ) {
     $kernel->post( $self->{sbox}->session_id(), 'submit', event => '_indices', job => 
        POE::Component::SmokeBox::Job->new(
	   ( $self->{backend} ? ( type => $self->{backend} ) : () ),
	   command => 'index',
        ),
     );
     return;
  }
  $kernel->yield( '_search' );
  return;
}

sub _indices {
  my ($kernel,$self,$data) = @_[KERNEL,OBJECT,ARG0];
  my ($result) = $data->{result}->results;
  unless ( $result->{status} == 0 ) {
     my $backend = $self->{backend} || 'CPANPLUS::YACSmoke';
     warn "There was a problem with the reindexing\n";
     return;
  }
  $kernel->yield( '_search' );
  return;
}

sub _search {
  my ($kernel,$self) = @_[KERNEL,OBJECT];
  if ( $self->{jobs} and ref $self->{jobs} eq 'ARRAY' ) {
     foreach my $distro ( @{ $self->{jobs} } ) {
        print "Submitting: $distro\n";
        $kernel->post( $self->{sbox}->session_id(), 'submit', event => '_smoke', job => 
           POE::Component::SmokeBox::Job->new(
	      ( $self->{backend} ? ( type => $self->{backend} ) : () ),
	      command => 'smoke',
	      module  => $distro,
           ),
        );
     }
  }
  if ( $self->{recent} ) {
    POE::Component::SmokeBox::Recent->recent( 
        url => $self->{url} || CPANURL,
        event => 'recent',
    );
  }
  if ( $self->{package} ) {
    warn "Doing a distro search, this may take a little while\n";
    POE::Component::SmokeBox::Dists->distro(
        event => 'dists',
        search => $self->{package},
        url => $self->{url} || CPANURL,
    );
  }
  if ( $self->{author} ) {
    warn "Doing an author search, this may take a little while\n";
    POE::Component::SmokeBox::Dists->author(
        event => 'dists',
        search => $self->{author},
        url => $self->{url} || CPANURL,
    );
  }
  if ( $self->{phalanx} ) {
    warn "Doing a phalanx search, this may take a little while\n";
    POE::Component::SmokeBox::Dists->phalanx(
        event => 'dists',
        url => $self->{url} || CPANURL,
    );
  }
  return if !$self->{recent} and ( $self->{package} or $self->{author} or $self->{phalanx} or ( $self->{jobs} and ref $self->{jobs} eq 'ARRAY' ) );
  POE::Component::SmokeBox::Recent->recent( 
      url => $self->{url} || CPANURL,
      event => 'recent',
  );
  return;
}

sub _submission {
  my ($kernel,$self,$state,$data) = @_[KERNEL,OBJECT,STATE,ARG0];
  if ( $data->{error} ) {
     warn $data->{error}, "\n";
     return;
  }
  if ( $state eq 'recent' and $self->{reverse} ) {
     @{ $data->{$state} } = reverse @{ $data->{$state} };
  }
  foreach my $distro ( @{ $data->{$state} } ) {
     print "Submitting: $distro\n";
     $kernel->post( $self->{sbox}->session_id(), 'submit', event => '_smoke', job => 
        POE::Component::SmokeBox::Job->new(
	   ( $self->{backend} ? ( type => $self->{backend} ) : () ),
	   command => 'smoke',
	   module  => $distro,
        ),
     );
  }
  return;
}

sub _smoke {
  my ($kernel,$self,$data) = @_[KERNEL,OBJECT,ARG0];
  my $dist = $data->{job}->module();
  my ($result) = $data->{result}->results;
  print "Distribution: '$dist' finished with status '$result->{status}'\n";
  my $run_time = $result->{end_time} - $result->{start_time};
  $self->{stats}->{max_run} = $run_time if $run_time > $self->{stats}->{max_run};
  $self->{stats}->{min_run} = $run_time if $self->{stats}->{min_run} == 0;
  $self->{stats}->{min_run} = $run_time if $run_time < $self->{stats}->{min_run};
  $self->{stats}->{_sum} += $run_time;
  $self->{stats}->{totaljobs}++;
  $self->{stats}->{avg_run} = $self->{stats}->{_sum} / $self->{stats}->{totaljobs};
  $self->{stats}->{idle}++ if $result->{idle_kill};
  $self->{stats}->{excess}++ if $result->{excess_kill};
  $self->{_jobs}--;
  return;
}

'smoke it!';
__END__

=head1 NAME

App::SmokeBox::Mini - the guts of the minismokebox command

=head1 SYNOPSIS

  #!/usr/bin/perl
  use strict;
  use warnings;
  BEGIN { eval "use Event;"; }
  use App::SmokeBox::Mini;
  App::SmokeBox::Mini->run();

=head2 run

This method is called by L<minismokebox> to do all the work.

=head1 AUTHOR

Chris C<BinGOs> Williams <chris@bingosnet.co.uk>

=head1 LICENSE

Copyright E<copy> Chris Williams

This module may be used, modified, and distributed under the same terms as Perl itself. Please see the license that came with your Perl distribution for details.

=cut


Makefile.PL
use strict;
use inc::Module::Install;
name 'minismokebox';
author 'Chris Williams <chris@bingosnet.co.uk>';
version_from 'lib/App/SmokeBox/Mini.pm';
abstract_from 'bin/minismokebox';
install_script 'bin/minismokebox';
license 'perl';
perl_version '5.006';
build_requires 'Test::More' => 0.47;
build_requires 'Test::POE::Server::TCP' => 0.12;
build_requires 'POE::Filter::HTTP::Parser' => 0.08;
build_requires 'HTTP::Response' => 0;
build_requires 'HTTP::Date' => 0;
requires 'Config::Tiny' => 2.12;
requires 'File::Spec' => 0;
requires 'Pod::Usage' => '1.00';
requires 'Time::Duration' => 1.06;
requires 'POE' => 1.0001;
requires 'POE::Component::SmokeBox' => 0.14;
requires 'POE::Component::SmokeBox::Recent' => 1.12;
requires 'POE::Component::SmokeBox::Dists' => 0.08;
if ( !( can_use('Event') and can_use('POE::XS::Queue::Array') ) and can_cc() ) {
print <<NOTICE;

\a--------------------------------------------------------------
Hey,

You appear to have a working compiler. I can include 'Event'
and 'POE::XS::Queue::Array' which will make POE work a lot 
faster.

--------------------------------------------------------------

NOTICE

my $prompt = prompt( 'Do you want to install \'Event\' and \'POE::XS::Queue::Array\' [Y/n]?', 'Y' );

if ( $prompt =~ /^Y/i ) {
  requires 'Event' => 1.11;
  requires 'POE::XS::Queue::Array' => 0.005;
}

}
clean_files '.smokebox jobs.txt';
WriteAll();
